gr-azure-software-radio/python/eventhub_sink.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the GNU General Public License v3.0 or later.
+# See License.txt in the project root for license information.
+#
+
+import json
+import pmt
+
+from gnuradio import gr
+from azure.eventhub import EventHubProducerClient, EventData
+from azure.core.credentials import AzureSasCredential
+from azure.identity import DefaultAzureCredential
+
+
+class EventHubSink(gr.sync_block):
+    """ Converts GNU Radio PMT to JSON and sends event to Azure Event Hub.
+
+    This block supports multiple ways to authenticate to Event Hub. Users can directly
+    specify either a connection string, an SAS token, or use credentials
+    supported by the DefaultAzureCredential class, such as environment variables.
+
+        Args:
+    Auth Method: Determines how to authenticate to Event Hub. May be one of
+        "connection_string", "sas", or "default".
+    Connection String: Event Hub connection string used for
+        authentication if Auth Method is "connection_string".
+    Shared Access Signature (SAS) token. This is required if using "default" or
+        "sas" authentication.
+    Event Hub Host Name:  The fully qualified host name for the Event Hub namespace. This is
+        required if using "default" or "sas" authentication.
+    Event Hub Name: The Event Hub must be created in Azure before trying to
+        send events to it.
+    Partition ID: The partition id to send events to.
+    """
+    # pylint: disable=too-many-arguments, no-member
+    def __init__(
+            self,
+            authentication_method: str = "default",
+            connection_str: str = None,
+            sas_token: str = None,
+            eventhub_host_name: str = None,
+            eventhub_name: str = None,
+            partition_id: str = None):
+
+        gr.sync_block.__init__(self,
+                               name="eventhub_sink",
+                               in_sig=[],
+                               out_sig=[])
+
+        self.partition_id = partition_id
+
+        self.eventhub_producer = get_eventhub_producer_client(
+            authentication_method=authentication_method,
+            eventhub_name=eventhub_name,
+            connection_str=connection_str,
+            sas_token=sas_token,
+            eventhub_host_name=eventhub_host_name
+        )
+
+        self.message_port_register_in(pmt.intern('in'))
+        self.set_msg_handler(pmt.intern('in'), self.handle_msg)
+
+    def handle_msg(self, msg):
+        """
+        This function handles the message passing operating (PMT message) from GNU Radio.
+        It converts the PMT structure to python and later to JSON to send as an event.
+        """
+        pmsg = pmt.to_python(msg)
+        jmsg = json.dumps(pmsg)
+        event_batch = self.eventhub_producer.create_batch(
+            partition_id=self.partition_id)
+        event_batch.add(EventData(jmsg))
+        self.eventhub_producer.send_batch(event_batch)
```

Comment:
any downside to using batch here? Batch is typically meant for N > 1 messages.

I think may just want to send event.

Reply:
API for the PublisherClient only supports "send_batch" https://azuresdkdocs.blob.core.windows.net/$web/python/azure-eventhub/latest/azure.eventhub.html